### DISCUSSION TITLE
Difference in saturation index calculated manualy and from AquousPropeties

### DISCUSSION THREAD

**Discussion by avdudchenko** [reaktoro/reaktoro#393] [General]
We have been working on integrating reaktoro into WaterTAP modeling platform. And one of the things we ended up having to do is to manually calculate the saturation index to properly propagate derivatives (as aqueous properties don't support sensitivity (https://github.com/reaktoro/reaktoro/discussions/380)  

One thing we noted is there are a small difference between manual calculation and the values reported by aqueous properties which I think stem from how we do pressure correction. 

Specifically for PhreeqC database with analytical function (but might be same result in other methods as well), the log_k is corrected via dV*(P-1)/(RT*2.303) - per (https://water.usgs.gov/water-resources/software/PHREEQC/documentation/phreeqc3-html/phreeqc3-36.htm) 

 This works well for some scalants like Calcite where differences could be very minor, but for Gypsum for example, this can result in larger differences (manual SI = 0.57 vs Aquous prop =0.58) 
 
How does Reaktoro handle pressure correction (as it does not seem to use the same approach to PhreeqC), and how would we replicate this without using Aquous properties calculations? 

Code the reproduces the issue:
```
import reaktoro as rkt

import json

import numpy as np

water_spec = {
    "Na+": 1120 / 22.989769e3,
    "Ca+2": 900 / 40.078e3,
    # "Cl-": 1750 / 35.453e3,
    "Sr+2": 3 / 85.62e3,
    "SO4-2": 1011 / 96.06e3,
    "HCO3-": 1200 / 61.0168e3,
    "H2O": 1000 / 18.01528,
}

temperature = 293.15
pressure = 100

scalant = "Gypsum"
db = rkt.PhreeqcDatabase("pitzer.dat")

species = " ".join(list(water_spec.keys()))
solution = rkt.AqueousPhase(rkt.speciate(species))
solution.set(rkt.ActivityModelPitzer())

system = rkt.ChemicalSystem(db, solution)
state = rkt.ChemicalState(system)

""" get props for reaction """
aprops = rkt.AqueousProps(state)
spec = aprops.saturationSpecies().get(scalant)
thermo_model = spec.standardThermoModel()

pr = spec.props(25, "C", 1, "atm")
calcite_volume = pr.V0
jsp = thermo_model.params().dumpJson()
jsp_dict = json.loads(jsp)
jsp = jsp_dict[0]["PhreeqcLgK"]
""" analytical log k calc for phreeqc"""
lgk = (
    jsp["A1"]
    + jsp["A2"] * temperature
    + jsp["A3"] * temperature**-1
    + jsp["A4"] * np.log10(temperature)
    + jsp["A5"] * temperature**-2
    + jsp["A6"] * temperature**2
)
""" getting dV """
dV = pr.V0
volume_reactants = 0
species = []

for s, mol in spec.reaction().reactants():
    spec = system.species().get(s.name())
    thermo_model = spec.standardThermoModel()
    pr = spec.props(25, "C", 1, "atm")
    volume_reactants += pr.V0 * abs(mol)
    species.append((s.name(), abs(mol)))
    print(s.name(), mol)
""" applying phreeqc pressure correction to logK """
lgk_p = lgk
lgk_p = lgk_p - (dV - (volume_reactants)) * (
    (pressure * 101325 - 101325) / (np.log(10) * 8.31446261815324 * temperature)
)

for ion, values in water_spec.items():
    state.set(ion, values, "mol")

state.set("H2O", 700 / 18.01528, "mol")
state.temperature(temperature, "kelvin")
state.pressure(pressure, "atm")
specs = rkt.EquilibriumSpecs(system)
""" Computing aquous props """
params = {}
specs.temperature()
specs.pressure()
specs.pH()
specs.charge()
specs.openTo("Cl-")
water_spec["H2O"] = 1000 / 18.01528
conditions = rkt.EquilibriumConditions(specs)
conditions.pH(7.07)
conditions.temperature(temperature, "kelvin")
conditions.pressure(pressure, "atm")
conditions.charge(0.0)
solver = rkt.EquilibriumSolver(specs)
result = solver.solve(state, conditions)
state.props().update(state)
manual_sat_index = (
    sum(
        float(coff * state.props().speciesActivityLn(specie).val() / np.log(10))
        for specie, coff in species
    )
    + lgk_p
)
aprops = rkt.AqueousProps(state)
aq_sat_index = aprops.saturationIndex(scalant)

print(f"manual calc {manual_sat_index}, aq_props_calc {aq_sat_index}")



```
Output :  manual calc 0.5197479615835636, aq_props_calc 0.522041770571369

**Comment by allanleal** [comment]
Thanks for reporting this. Did you check if relative differences are the same across other minerals? For the reported Gypsum SI values above, the difference is about 0.5%. Reaktoro uses the same pressure correction approach used in PHREEQC when a PHREEQC database is used. 

I would also use the same constants adopted in Reaktoro. For example, molar mass of H2O (taken from the Species object for water), replace 2.303 by log(10), check if same universal gas constant value, etc. If I'm not mistaken, the use of pitzer.dat leads to water molar mass equal to 18.016 instead of 18.01528. But maybe this is not relevant in this case.

I'm answering from my phone and cannot run your code at the moment. Currently out of the office. I'll be back next week and could check more carefully.

**Reply by avdudchenko** [reply]
Thanks.  I think I found the error. I was not multiplying the reaction volume by the moles in the reaction resulting in an incorrect total reaction volume.  This significantly reduces the difference now. However, a small error remains. I think this mostly resolves the issue then, as I imagine the error is due to rounding at this stage. I would appreciate a second look when you get back. 

Gypsum:
at 100atm - Manual SI = 0.519747 AquousProps = 0.522041 
Calcite:
at 100atm - Manual SI = 1.185988 AquousProps = 1.1875273
Celestite:
at 100atm - Manual SI = -0.45402 AquousProps = -0.45079
SiO2:
at 100atm - Manual SI = -0.47705 AquousProps = -0.47382
